src/nexmark/queries/q15.rs

```diff
@@ -74,8 +69,99 @@ pub struct Q15Output {
     rank3_auctions: usize,
 }
 
+type OrdinalDate = (i32, u16);
+
 type Q15Stream = Stream<Circuit<()>, OrdZSet<Q15Output, isize>>;
 
+/*
+fn dump_tuples<B>(batch: &B)
+where
+    B: BatchReader<Time=()>,
+    B::Key: std::fmt::Debug,
+    B::Val: std::fmt::Debug,
+    B::R: std::fmt::Debug,
+{
+    let mut cursor = batch.cursor();
+
+    while cursor.key_valid() {
+        while cursor.val_valid() {
+            let w = cursor.weight();
+            println!("{:?} -> {:?} x {:?}", cursor.key(), cursor.val(), w);
+            cursor.step_val();
+        }
+        cursor.step_key();
+    }
+}
+*/
+
+impl<P, K, V> Stream<Circuit<P>, OrdIndexedZSet<K, V, isize>>
+where
+    P: Clone + 'static,
+    K: Default + DeepSizeOf + Clone + Ord + Hash + Send + 'static,
+    V: Default + DeepSizeOf + Clone + Ord + Hash + Send + 'static,
+{
+    /// Outer join:
+    /// - returns the output of `join_func` for common keys.
+    /// - returns the output of `left_func` for keys only found in `self`, but
+    ///   not `other`.
+    /// - returns the output of `right_func` for keys only found in `other`, but
+    ///   not `self`.
+    fn outer_join<V2, F, FL, FR, O>(
+        &self,
+        other: &Stream<Circuit<P>, OrdIndexedZSet<K, V2, isize>>,
+        join_func: F,
+        left_func: FL,
+        right_func: FR,
+    ) -> Stream<Circuit<P>, OrdZSet<O, isize>>
+    where
+        Self: FilterMap<Circuit<P>, R = isize>,
+        V2: Default + DeepSizeOf + Clone + Ord + Hash + Send + 'static,
+        O: Clone + Default + Ord + DeepSizeOf + 'static,
+        F: Fn(&K, &V, &V2) -> O + Clone + 'static,
+        for<'a> FL: Fn(<Self as FilterMap<Circuit<P>>>::ItemRef<'a>) -> O + Clone + 'static,
+        for<'a> FR: Fn(
+                <Stream<Circuit<P>, OrdIndexedZSet<K, V2, isize>> as FilterMap<Circuit<P>>>::ItemRef<
+                    'a,
+                >,
+            ) -> O
+            + Clone
+            + 'static,
+    {
+        let center = self.join_generic::<(), _, _, _, _>(other, move |k, v1, v2| {
+            std::iter::once((join_func(k, v1, v2), ()))
+        });
+        let left = self
+            .antijoin::<(), _>(other)
+            .map_generic::<_, _, _>(move |x| left_func(x));
+        let right = other
+            .antijoin::<(), _>(self)
+            .map_generic::<_, _, _>(move |x| right_func(x));
```

Comment:
I think a lot of this is redundant
```suggestion
            .map_generic(left_func);
        let right = other
            .antijoin::<(), _>(self)
            .map_generic(right_func);
```